core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -2035,6 +2040,107 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
+  }
+
+  test("Test dagScheduler.shouldUnregisterMapOutput with map stage not running") {
+    reInit()
+    val shuffleMapRdd = new MyRDD(sc, 2, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, new HashPartitioner(2))
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 2, List(shuffleDep), tracker = mapOutputTracker)
+    submit(reduceRdd, Array(0, 1))
+    complete(taskSets(0), Seq(
+      (Success, makeMapStatus("hostA", reduceRdd.partitions.length)),
+      (Success, makeMapStatus("hostB", reduceRdd.partitions.length))))
+    // The MapOutputTracker should know about both map output locations.
+    assert(mapOutputTracker.getMapSizesByExecutorId(shuffleId, 0).map(_._1.host).toSet ===
+      HashSet("hostA", "hostB"))
+
+    // The first result task fails, with a fetch failure for the output from the first mapper.
+    runEvent(makeCompletionEvent(
+      taskSets(1).tasks(0),
+      FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, 0, "ignored"),
+      null))
+    assert(sparkListener.failedStages.contains(1))
+
+    val mapStatuses = mapOutputTracker.shuffleStatuses(shuffleId).mapStatuses
+    // unregisterMapOutput with a fetchFailed.
+    assert(mapStatuses.count(_ != null) === 1)
+    assert(mapStatuses(1).location === makeBlockManagerId("hostB"))
+  }
+
+  test("Test dagScheduler.shouldUnregisterMapOutput with map stage re-running") {
+    reInit()
+    val shuffleMapRdd = new MyRDD(sc, 3, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, new HashPartitioner(2))
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 2, List(shuffleDep), tracker = mapOutputTracker)
+    submit(reduceRdd, Array(0, 1))
+    complete(taskSets(0), Seq(
+      (Success, makeMapStatus("hostA", reduceRdd.partitions.length)),
+      (Success, makeMapStatus("hostB", reduceRdd.partitions.length)),
+      (Success, makeMapStatus("hostC", reduceRdd.partitions.length))))
+    // The MapOutputTracker should know about both map output locations.
+    assert(mapOutputTracker.getMapSizesByExecutorId(shuffleId, 0).map(_._1.host).toSet ===
+      HashSet("hostA", "hostB", "hostC"))
+
+    runEvent(makeCompletionEvent(
+      taskSets(1).tasks(0),
+      FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, 0, "ignored"),
+      null))
+    runEvent(makeCompletionEvent(
+      taskSets(1).tasks(1),
+      FetchFailed(makeBlockManagerId("hostB"), shuffleId, 1, 1, 0, "ignored"),
+      null))
+
+    assert(sparkListener.failedStages.contains(1))
+
+    // Wait for a long time to make sure the map stage was resubmitted.
+    eventually(timeout(1000 milliseconds), interval(10 milliseconds)) {
+      assert(scheduler.runningStages.nonEmpty && taskSets.size == 3)
+    }
+
+    runEvent(makeCompletionEvent(
+      taskSets(2).tasks(0),
+      Success,
+      makeMapStatus("hostA", reduceRdd.partitions.length)))
+
+    runEvent(makeCompletionEvent(
+      taskSets(1).tasks(1),
```

Comment:
`taskSets(1).tasks(1)` already completed with `FetchFailure` above, why it can complete again?

Reply:
Oh, I'm sorry,i have update my ut, thanks.